supertokens_python/recipe/session/recipe_implementation.py

```diff
@@ -46,10 +46,93 @@
 if TYPE_CHECKING:
     from typing import List, Union
     from supertokens_python import AppInfo
-    from supertokens_python.querier import Querier
 
-from .constants import JWKCacheMaxAgeInMs, JWKRequestCooldownInMs
 from .interfaces import SessionContainer
+from supertokens_python.querier import Querier
+
+
+from typing_extensions import TypedDict
+from os import environ
+
+
+class JWKSConfigType(TypedDict):
+    cache_max_age: int
+    refresh_rate_limit: int
+
+
+JWKSConfig: JWKSConfigType = {
+    "cache_max_age": 6000,
+    "refresh_rate_limit": 500,
+}
+
+cached_jwk_client: Optional[JWKClient] = None
+mutex = RWMutex()
+
+# only for testing purposes
+def reset_jwks_cache():
+    with RWLockContext(mutex, read=False):
+        global cached_jwk_client
+        cached_jwk_client = None
+
+
+def get_jwk_client_from_cache() -> Optional[JWKClient]:
+    with RWLockContext(mutex, read=True):
+        if cached_jwk_client is not None:
+            # This means that we have valid JWKs for the given core path
+            # We check if we need to refresh before returning
+
+            # This means that the value in cache is not expired, in this case we return the cached value
+            # Note that this also means that the SDK will not try to query any other Core (if there are multiple)
+            # if it has a valid cache entry from one of the core URLs. It will only attempt to fetch
+            # from the cores again after the entry in the cache is expired
+            if cached_jwk_client.is_fresh():
+                if environ.get("SUPERTOKENS_ENV") == "testing":
+                    log_debug_message("Returning JWKS from cache")
+                return cached_jwk_client
+
+        return None
+
+
+def find_jwk_client() -> JWKClient:
```

Comment:
this should take a kid and return the right key for that kid:
- if the cache is none, or has expired, we fetch
- if the kid in the cache or fetched value is missing, we fetch
- we update the cache if needed (if we fetch)
- return the right key.
- if no key found, then throw error. Which sohuld translate to try refresh token

Reply:
Done.